Plugin 092: DL-bus from Technische Alternative (PR2)

uwekaditz opened the pull request:
Removed all commits I merged in the mega branch

uwekaditz committed:
[P092] New PlugIn to decode the DL-Bus

The DL-Bus is used in heating control units e.g. sold by Technische Alternative (www.ta.co.at).
The idea for this plug-in is based on Martin Kropf's project UVR31_RF24 (https://github.com/martinkropf/UVR31_RF24)
The plug-in is tested and workis fine for the ESR21 device.
The plug-in should be also able to decode the information from the UVR31, UVR1611 and UVR 61-3 devices but not tested yet.
The selected input needs a voltage divider as follows because the DL-Bus runs on 12 volts.

uwekaditz committed:
[P092] New PlugIn to decode the DL-Bus

Added to PLUGIN_SET_STABLE
Basic documentation added

uwekaditz committed:
Update _P092_DLbus.ino

uwekaditz committed:
Merge branch 'UweKaditz' of https://github.com/uwekaditz/ESPEasy into UweKaditz

uwekaditz committed:
Update _P092_DLbus.ino

Removed all commits I merged in the mega branch

uwekaditz committed:
[P092] Timing optimized

Removed the while (P092_receiving) loop.
Starting of the receiving and processing of the received bit stream iare now done in the PLUGIN_ONCE_A_SECOND call
PLUGIN_READ call just uses the already processed data

uwekaditz commented:
I just finished my first test with optimized timing.
Attached are a snap shot of the timing and a part of the log data.
[DLbus.log](https://github.com/letscontrolit/ESPEasy/files/3964380/DLbus.log)
![Timing DLbus calls](https://user-images.githubusercontent.com/55511593/70855045-8d397e00-1ec4-11ea-94d5-99020d352286.png)

uwekaditz commented:
Should I close the old PR for the DLbus?
The load is also acceptable 1(16%).
![WemosMain](https://user-images.githubusercontent.com/55511593/70855098-57e16000-1ec5-11ea-8160-0d7394f50cce.JPG)
I use following device settings (refreshing every 60 seconds):
![Devices](https://user-images.githubusercontent.com/55511593/70855112-96771a80-1ec5-11ea-80d5-da80d63f8fcd.JPG)

uwekaditz committed:
[P092] Memory usage optimized

CHG: Moved the array for the received bit changes to stativ uint_8t, the ISR call uses only a volatile pointer to it
CHG: some more defines and name changes for better explanation

TD-er commented on src/_P092_DLbus.ino:
```diff
@@ -798,8 +807,8 @@ boolean Plugin_092_CheckTimings(void) {
 void P092_Pin_changed(void) {
-  uint32_t DLbus_time_diff = usecPassedSince(DLbus_last_bit_change);  // time difference to previous pulse in µs
-  DLbus_last_bit_change = micros();                                   // save last pin change time
+  uint32_t DLbus_time_diff = usecPassedSince(DLbus_TimeLastBitChange);  // time difference to previous pulse in µs
```
time diff is a signed value.
So when some timer has not yet passed, the value will be negative, or very large when stored in an unsigned value.

uwekaditz replied:
Changed

TD-er commented on src/_P092_DLbus.ino:
```diff
@@ -850,21 +873,22 @@ String log;
       return false;
     }
-    if ((DLbus_bit_number-DLbus_start_bit)<(DLbus_pulse_number/4)) {
+    if ((DLbus_bit_number-DLbus_start_bit)<((DLbus_pulse_number - DLBus_ReserveBytes)/DLBus_BitChangeFactor)) {
```
Maybe you can create a variable here describing what you try to compare here?
That makes the code more readable.
Also it prevents computing the same value twice (for the log)

uwekaditz replied:
Changed

TD-er commented on src/_P092_DLbus.ino:
```diff
@@ -809,10 +818,23 @@ void P092_Pin_changed(void) {
+      // check if sync is received
+      if (val & DLbus_FlagLongerThanTwiceDoubleWidth) {
+        // sync received
+        *DLbus_PtrChangeBitStream = !(val & 0x01);
```
There is no guarantee this pointer is valid when this function is called as the pointer is not guaranteed to be set when the interrupt is attached.
Also, why do you use a pointer to an object which is always present and never increase the pointer?
It would make sense to use a pointer when you use it as an iterator through the array, or when the array has to be created at init for example.

uwekaditz replied:
I use this pointer as volatile because it is called in the ISR call.
The array the pointer points to is not longer volatile, that means not stored in th IRAM section.
This pointer is valid because it is set before the `P092_receiving` is set tu `true` (line 700)!

TD-er replied:
Not 100% sure how the compiler will use the array here.
But I understand now why you did it.

I missed the `P092_receiving` in line 700, but you are right. This then acts as a guard for the initialized pointer.

uwekaditz replied:
It's line 701 in the changed file.

uwekaditz replied:
Is there a chance to see the assembler code created by the compiler?

TD-er replied:
I'm sure there is, but even then I don't know how to interpret the allocated memory ranges to see if it will cause an issue or not.
The core libs have some checks for missing attributes for ISR related code, but I am not sure if it also checks (or need to check) for memory access address ranges.
I guess you need to check stability over time.

uwekaditz replied:
So far the changed code runs stable for about 90 minutes.
There are no sync information of the DLbus missed.
The received bit changes can be decoded (manchester code) and the CRC check is always passed.
```
4507343: PLUGIN_092_READ
4507344: Get Sensor1: 21.7
4507439: PLUGIN_092_READ
4507439: Get Sensor2: 29.2
4507480: PLUGIN_092_READ
4507480: Get DigitalOutput1: 0.0
4507521: PLUGIN_092_READ
4507522: Get AnalogOutput1: 0.0
4508736: WD : Uptime 75 ConnectFailures 0 FreeMem 18192 WiFiStatus 3
4512499: SYS : 75.00,15.30,18296.00,3376.00
4519265: P092_receiving ...
4521266: Receive stopped.
4521267: DLbus_ChangeBitStream copied.
4521267: Processing...
4521268: Invert bit stream...
4521269: StartBit: 36 / EndBit: 413
4521271: Check CRC...
4521271: Received data OK
4538736: WD : Uptime 76 ConnectFailures 0 FreeMem 18192 WiFiStatus 3
4552266: P092_receiving ...
4554266: Receive stopped.
4554267: DLbus_ChangeBitStream copied.
4554267: Processing...
4554268: Invert bit stream...
4554269: StartBit: 36 / EndBit: 414
4554271: Check CRC...
4554271: Received data OK
4567366: PLUGIN_092_READ
4567367: Get Sensor1: 21.7
4567488: PLUGIN_092_READ
4567489: Get Sensor2: 29.0
4567538: PLUGIN_092_READ
4567538: Get DigitalOutput1: 0.0
4567578: PLUGIN_092_READ
4567578: Get AnalogOutput1: 0.0
4568743: WD : Uptime 76 ConnectFailures 0 FreeMem 16568 WiFiStatus 3
```

TD-er replied:
Well if there is an issue with ISR code, caused by variables not defined like they should, then it will simply crash with some kind of exception.

uwekaditz committed:
[P092] Changes requested by TDer

CHG: variable to calculate RequiredBitStreamLength
CHG: function usecPassedSince() is long

uwekaditz committed:
[P092] Documentation added

NEW: documentation

uwekaditz commented:
So far the changed code runs stable for almost one day.
![main](https://user-images.githubusercontent.com/55511593/70902337-d2f96200-1ffc-11ea-9c2f-7e50a4939e7b.JPG)

uwekaditz commented:
I got some more question:
1. Is there a way to limit the global value `Interval`?
2. Is there a way to avoid the memory overflow error for the travis build ( e.g. for `test_ESP8266_4M1M_VCC_MDNS_SD`?

TD-er commented:
About the build sizes. That's something I have to work on.
We keep hitting that limit, so we must make different build selections.

> Is there a way to limit the global value Interval?

I will have a look at it.

uwekaditz committed:
[P092] More documentation added

NEW: docs for ESR21 and UVR31 units

uwekaditz commented:
So far the changed code runs stable for more than 4 days.
I don't think there is an issue with the ISR code or variables.
![main](https://user-images.githubusercontent.com/55511593/71209859-e03c7800-22ac-11ea-90b3-c70eddba04e0.JPG)

uwekaditz commented:
Is there a chance to test the layout of the rst files used for documentation?

TD-er commented:
> Is there a chance to test the layout of the rst files used for documentation?

Yes, you can use the Python environment used in PlatformIO.
So from within your IDE (Atom or VScode), you open a terminal (PIO menu or PIO icon bar) and it will activate your Python virtual env.
If the ESPEasy project has been loaded, it will open it in the ESPEasy dir.

Then cd to the docs dir.
Just to be sure (if never run before), install the needed packages using pip:

```
pip install -r requirements.txt
```

then run the `make html` command, or `.\make.bat html` in Windows.

The generated docs will be in the docs/build dir.

uwekaditz commented:
Thanks for the info.
I installed everything and run the `.\make.bat html` in Windows.
I got following error message
![error1](https://user-images.githubusercontent.com/55511593/71288068-5652e480-236a-11ea-9a68-e68c90c59b7d.JPG)
I removed the C014 files but now I get the same message for `Plugin/P000_Buzzer_RTTTL`
![error](https://user-images.githubusercontent.com/55511593/71287990-1429a300-236a-11ea-8a40-256136456da0.JPG)
**What is wrong?**

uwekaditz committed:
[P092] Documentation updated

NEW: More devices
CHG: can be compiled by sphinx without warnings or errors

TD-er commented:
Looks like there is some non-ASCII character in the file?

uwekaditz commented:
I can't see any 0x81 character in the files `C014.rst` or `P000_Buzzer_RTTTL.rst`
I opened both files in a hex editor.

uwekaditz commented:
There is also no warning while compiling. I got a few in my files using the ² character.
Attached is the log of the last make.
[sphinx-err-38rxo_4m.log](https://github.com/letscontrolit/ESPEasy/files/3990209/sphinx-err-38rxo_4m.log)

TD-er commented:
I just checked the docs dir from a Linux shell:
```
gijsnoorlander@penguin:~/GitHub/TD-er/ESPEasy/docs$ grep --color='auto' -P -n "[^\x00-\x7f]" * -R 
```
Apparently we have a lot more files with non-ASCII characters in it.

uwekaditz committed:
[P093] Documentation updated

uwekaditz committed:
Merge tag 'mega-20200204' into UweKaditz

uwekaditz commented:
Any more changes requested for this pull request?

TD-er commented on src/_P092_DLbus.ino:
```diff
@@ -0,0 +1,1288 @@
+  uint8_t       IdxkWh3;
+  uint8_t       IdxMWh3;
+  uint8_t       IdxCRC;
+} P092_DataSettings;
```
This one and these:
```c++
uint8_t P092_MaxIdx[P092_DLbus_OptionCount];
int P092_ValueType[P092_DLbus_ValueCount];
int P092_ValueIdx[P092_DLbus_ValueCount];
```
are all taking resources (memory) even when the plugin is not active.
Also they make it impossible to have multiple instances of the same plugin to be active, unless they all are allowed to share the same objects.

Please have a look at the plugins using a struct that's inheriting from `PluginTaskData_base` (e.g. the GPS plugin)
The idea is to have only a pointer allocated from a vector of task data, which is then cast to the right type if it exists.
This way a task has its own set of data and thus allowing multiple instances of the same plugin, but also it is not taking any memory when not selected and enabled.
The way you have it now, does take quite a lot of memory, even when it is not used but only included in a build.

uwekaditz replied:
I understand your concern about the limited ressources on the ESP82 devices.
Nevertheless I need to explain, why `PluginTaskData_base ` is not working for this plug_in.
1. Only one controller device type can be  connected to a DL bus instance (ESR21, UVR31, UVR1611 or UVR 61-3).
2.  The data read and decoded from the DL bus are global for all task using the P092_DLbus-plug-in because they are read only by the global interrupt function of the IO pin (CONFIG_PIN1). The decoding happens also only once after all bytes are received.
3. A new device task must be created for each sensor value to be extracted from the DL bus (using the same device type and IO pin!). This new task only extracts the data for the requested sensor value from the received data (golabl, see 2.). I had to program this restriction because up to 37 sensor values can be combined in one DL bus frame (depending on the controller device type) and the ValueCount of any device task is limited to 4 (if I'm not wrong).

It would be very helpful for me if you have some example
- how to use the same pointer to allocated memory from different device tasks of the same type and
- how to allocate the memory to this pointer once the first device tasks is created and
- how to free the memory if the last device tasks of the same type is deleted.

I need to test the other changes I already made on my real system before I can upload the new source code.

TD-er replied:
You can also create a single struct (e.g. `p092_datastruct`) for this shared data and then just define a "global" pointer to this (initialized to `null_ptr`).
As soon as you init the first task for this, you will create a `new p092_datastruct;` and assign it to this pointer.
As long as the pointer is still a `null_ptr`, then the task is not initialized (and also not taking memory).
If you really need it, you will use resources and that's fine then.

uwekaditz replied:
See my last changes, I put as much as possible in a class, only created while using the P092 plug in.

TD-er commented on src/_P092_DLbus.ino:
```diff
@@ -0,0 +1,1288 @@
+            P092_MaxIdx[4] = 1; // Speed step
+            P092_MaxIdx[5] = 1; // Analog output
+            P092_MaxIdx[6] = 1; // Heat power (kW)
+            P092_MaxIdx[7] = 1; // Heat meter (MWh)
```
This switch statement does set all values for `P092_MaxIdx` in each case, while most have the same value as used in other cases.
So better set a default config before the switch statement and only the values that differ from the default then have to be set in each case.
Not only does it make more clear what differs among cases, but also the resulting binary is smaller.

You can also create defines for the indices of this array so you don't need to repeat comments, but have it commented in the code itself.

uwekaditz replied:
Done in the next version of _P092_DLbus.ino

TD-er commented on src/_P092_DLbus.ino:
```diff
@@ -0,0 +1,1288 @@
+  };
+  const String plugin_092_Indizes[P092_DLbus_ValueCount] = {
+    F("p092_Idx")
+  };
```
I really don't understand why these `plugin_092_Values` and `plugin_092_Indizes` have to be an array.
Their values are never initialized to anything other than the default constructor (see the `const` to prove it) and all values are the same, or at least the first one is as defined, not sure if the rest of the array is initialized with anything other than an empty String.
This does take quite a bit of stack allocation, which is unneeded and we already have so little stack available.

uwekaditz replied:
Changed in the next version of _P092_DLbus.ino

TD-er commented on src/_P092_DLbus.ino:
```diff
@@ -0,0 +1,1288 @@
+
+        for (int i = 0; i < P092_DLbus_ValueCount; i++) {
+          P092_ValueType[i] = PCONFIG(i + 1) >> 8;
+          P092_ValueIdx[i] = PCONFIG(i + 1) & 0x00FF;
```
`P092_ValueType` and `P092_ValueIdx` are only addressed for the same index, so there is no need to have them defined as an array.
You can define them here in the loop as single variables (even `const` values) and have them destructed at the end of the loop.
Now they are:
- allocated as an array (not needed at all) 
- allocated on the stack as array (try to prevent stack array declarations as much as possible, we really have very little stack)
- kept alive outside the for loop (not needed)
- Allocated as global variables, and thus taking resources even when the plugin is not selected or enabled.

uwekaditz replied:
Changed in the next version of _P092_DLbus.ino

TD-er commented on src/_P092_DLbus.ino:
```diff
@@ -0,0 +1,1288 @@
+boolean Plugin_092(uint8_t function, struct EventStruct *event, String& string)
+{
+  boolean success = false;
+  int OptionIdx, CurIdx;
```
Everywhere you use `CurIdx`, it is initialized with a value, so no need to have it allocated outside the scope of where you use it.
So allocate it on the stack where you assign it, then it is no longer taking resources when outside the scope where it is needed.
Probably the same with `OptionIdx`, have not checked that one.

TD-er commented on src/_P092_DLbus.ino:
```diff
@@ -0,0 +1,1288 @@
+      P092_DataSettings.IdxCRC = 30;
+
+      P092_OptionValueDecimals[6] = 1; //[0,1kW]     F("Heat power (kW)")
+      break;
```
Same here as with the other switch statement. Perhaps there is a proper default setting possible so you only have to set the values that differ.  (although not entirely sure it will be more clear here what the exact config per item is then as there are lots of values to be set here in this struct)

uwekaditz replied:
Done in the next version of _P092_DLbus.ino

TD-er commented on src/_P092_DLbus.ino:
```diff
@@ -0,0 +1,1288 @@
+  P092_DataSettings.IdxMWh2                 = P092_DataSettings.IdxkWh2 + 2;
+  P092_DataSettings.IdxHeatMeter3           = P092_DataSettings.IdxMWh2 + P092_DataSettings.MWhBytes;
+  P092_DataSettings.IdxkWh3                 = P092_DataSettings.IdxHeatMeter3 + P092_DataSettings.CurrentHmBytes;
+  P092_DataSettings.IdxMWh3                 = P092_DataSettings.IdxkWh3 + 2;
```
Why not create functions to get the derived values?
In the struct you can create get functions for these.
It will also make sure derived values are always in sync with the source values.
When using it like this, it is possible a source value is updated and the derived value is not, or the other way around.

If the relation to a derived value and source values is always the same, please use a function to get those derived values.

N.B. a get function like that will not increase the size of the struct you define it for.

uwekaditz replied:
Changed in the next version of _P092_DLbus.ino

TD-er commented on src/_P092_DLbus.ino:
```diff
@@ -0,0 +1,1288 @@
+uint16_t DLbus_bit_number;                                  // bit number of the received DLbus_ChangeBitStream
+
+uint8_t DLbus_WrongTimeCnt;
+uint16_t DLbus_WrongTimings[5][6];
```
And another 60 bytes allocated for `DLbus_WrongTimings` on the stack.  (sys stack, not cont stack)
This is allocated regardless whether the plugin is being used.
The same applies for `DLbus_ByteStream`, which is now already close to 200 bytes.

uwekaditz replied:
Changed in the next version of _P092_DLbus.ino

TD-er reviewed (changes requested):
Just had a quick look at the code and found quite a lot of allocations which do take a lot of resources regardless whether the plugin is actually used.

uwekaditz committed:
Memory improvement for _P092 plug in

Moved arrays into the class DLBus (only created while using the plug in)
Moved other arrays into function Plugin_092()

uwekaditz committed:
Merge branch 'mega' into UweKaditz

uwekaditz committed:
Merge branch 'UweKaditz' of https://github.com/uwekaditz/ESPEasy into UweKaditz

uwekaditz committed:
Merge conflicts solved

uwekaditz committed:
merged with release mega-20200305

merged with release mega-20200305

uwekaditz committed:
Merge branch 'UweKaditz' of https://github.com/uwekaditz/ESPEasy into UweKaditz

uwekaditz committed:
Merge branch 'UweKaditz' of https://github.com/uwekaditz/ESPEasy into UweKaditz

uwekaditz committed:
Merge branch 'mega' into UweKaditz

uwekaditz committed:
Merge branch 'mega' into UweKaditz

uwekaditz committed:
Merge branch 'mega' into UweKaditz

TD-er committed:
Merge branch 'mega' into UweKaditz

TD-er committed:
Merge branch 'mega' into UweKaditz

TD-er committed:
Merge branch 'mega' into UweKaditz

TD-er committed:
[P092] Mark plugin as  [TESTING]
